src/darwin/Framework/CHIP/CHIPDeviceController.mm

```diff
@@ -355,6 +357,47 @@ - (BOOL)stopDevicePairing:(uint64_t)deviceID error:(NSError * __autoreleasing *)
     return success;
 }
 
+- (BOOL)isDevicePaired:(uint64_t)deviceID error:(NSError * __autoreleasing *)error
+{
+    __block BOOL paired = NO;
+    if (![self isRunning]) {
+        [self checkForError:CHIP_ERROR_INCORRECT_STATE logMsg:kErrorNotRunning error:error];
+        return paired;
+    }
+    dispatch_sync(_chipWorkQueue, ^{
+        if ([self isRunning]) {
+            paired = self.cppCommissioner->DoesDevicePairingExist(chip::PeerId().SetNodeId(deviceID));
+        }
+    });
+
+    return paired;
+}
+
+- (BOOL)getConnectedDevice:(uint64_t)deviceID
+         completionHandler:(CHIPDeviceConnectionCallback)completionHandler
+                     queue:(dispatch_queue_t)queue
+                     error:(NSError * __autoreleasing *)error
+{
+    __block BOOL didSucceed = NO;
+    __block CHIP_ERROR errorCode = CHIP_ERROR_INCORRECT_STATE;
+    if (![self isRunning]) {
+        [self checkForError:errorCode logMsg:kErrorNotRunning error:error];
+        return didSucceed;
+    }
+
+    dispatch_sync(_chipWorkQueue, ^{
```

Comment:
Does this need to be a sync dispatch?  Couldn't it just communicate failures via the completion handler?

Reply:
The following would require extra handling in the user of `getConnectedDevice`
1. The user calls `getConnectedDevice`
2. The call to `connectionBridge->connect` returns error.
3. Before returning error, the underlying `DeviceController::GetConnectedDevice` has already called the callback (triggering completionHandler) due to error.
4. The `getConnectedDevice` returns `YES` after the `completionHandler` had already been called.

The user has to maintain a state that the completionHandler returned an error, so ignore the value returned by `getConnectedDevice`.

Reply:
Thinking more about it, it might be unlikely that `async` block runs before `getConnectedDevice()` actually returns. Thoughts?

Reply:
Updated the code to use async block

Reply:
Hmm.  So the right way for the user to use the return value of `getConnectedDevice` seems to be that if it's NO then the completion handler will definitely not get called.  And if it's YES then it can get called, or maybe was already called.

That last seems pretty dubious to me, and really hard to use; I'd think we would want to queue the completion handler on the caller's queue or something instead of just blindly calling into them on whatever thread.  @pan-apple, am I missing something or should we file a followup to do something like that?